### PR TITLE
Kokkos port for GRANULAR

### PR DESCRIPTION
**Pull request by valleymouth**
## Purpose

This pull request allows the use of GRANULAR's pair gran/hooke/history and fix freeze with Kokkos.

## Author(s)

Denis Taniguchi (Newcastle University)

## Backward Compatibility

This pull request doesn't break backward compatibility.

## Implementation Notes

Implementation was based on the bench/in.chute case, i.e. all the classes involved in running that case were ported.
Validation of the porting was also performed based on the bench/in.chute case. Running that case using Kokkos or not doesn't give exact same results but the overall behaviour is maintained (see attached log files).
AtomVecSphereKokkos was implemented to include radius and rmass atom attributes. NPairKokkos was modified to take into account size. FixNeighHistoryKokkos was also ported to account for the displacement history.
We still need to profile the code and check for simple optimisations opportunities, such as shear history memory access pattern. Since the original GRANULAR code relied exclusively on half neighbour list, the same was made in this implementation. Maybe implementing full list support could be an interesting thing to avoid atomic operations, which could lead to better performance specially on old GPUs.  

## Post Submission Checklist

_Please check the fields below as they are completed_
- [x] The feature or features in this pull request is complete
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] One or more example input decks are included
- [x] The source code follows the LAMMPS formatting guidelines

## Further Information, Files, and Links

[chute-regular.txt](https://github.com/lammps/lammps/files/2166174/chute-regular.txt)
[chute-kokkos.txt](https://github.com/lammps/lammps/files/2166175/chute-kokkos.txt)

### CODE REVIEW

**Comment by akohlmey** [comment]
@valleymouth thanks for your contribution, but please note, that your pull request fails to compile with KOKKOS enabled for OpenMP.

File: src/KOKKOS/Install.sh [hunk 1]
```diff
@@ -95,8 +97,14 @@ action fix_deform_kokkos.cpp
 action fix_deform_kokkos.h
 action fix_eos_table_rx_kokkos.cpp fix_eos_table_rx.cpp
 action fix_eos_table_rx_kokkos.h fix_eos_table_rx.h  
+action fix_freeze_kokkos.cpp
```
Comment (stanmoore1) [inline src/KOKKOS/Install.sh]: This and similar files need to depend on the GRANULAR file.

**Comment by stanmoore1** [comment]
@valleymouth thanks for all the work on this. In general it looks good.

**Comment by stanmoore1** [comment]
@valleymouth I'm running this through our internal Kokkos regression tests.

**Comment by stanmoore1** [comment]
@valleymouth the Kokkos OpenMP regression tests had some failures. Some were "invalid free", some were segfaults, and some were LAMMPS errors. Here are the failing tests:

```
[examples/bench', 'chute']
[examples/srd', 'srd.mixture']
[examples/pour', 'pour.2d.molecule']
[examples/pour', 'pour']
[examples/pour', 'pour.2d']
[examples/colloid', 'colloid']
[examples/ASPHERE/dimer', 'dimer.mp']
[examples/ASPHERE/dimer', 'dimer']
[examples/ASPHERE/poly', 'poly.mp']
[examples/ASPHERE/poly', 'poly']
[examples/ASPHERE/ellipsoid', 'ellipsoid.mp']
[examples/ASPHERE/ellipsoid', 'ellipsoid']
[examples/ASPHERE/box', 'box.mp']
[examples/ASPHERE/box', 'box']
[examples/ASPHERE/star', 'star.mp']
[examples/ASPHERE/star', 'star']
```

**Comment by stanmoore1** [comment]
OK all the OpenMP tests now seem to be running expect `pour`. @valleymouth can you take a look at that? I'm going to look at GPU now.

**Comment by stanmoore1** [comment]
@valleymouth if necessary you can put in a meaningful error message for fix pour if it can't work correctly (no need to port fix pour to Kokkos right now).

**Comment by valleymouth** [comment]
@stanmoore1 Ok, I will look into fix pour problem.

**Comment by stanmoore1** [comment]
@valleymouth I restarted the regression tests again. Apparently Kokkos doesn't appreciate when a View is declared randomread but not initialized. I'm going to submit a Kokkos issue about that.

**Comment by stanmoore1** [comment]
I submitted a Kokkos issue: https://github.com/kokkos/kokkos/issues/1691

**Comment by stanmoore1** [comment]
It looks like there is a bug in either the pack/unpack border or pack/unpack exchange routines in AtomVecSphere. If I run the `colloid` example with `-pk kokkos newton on neigh half` then it crashes with a segfault, but if I force the exchange and border communication to run in non-Kokkos mode, then it runs fine: `-pk kokkos newton on neigh half comm/exchange no`.

**Comment by stanmoore1** [comment]
This for OpenMP, so it is more than just a missing sync/modify.

**Comment by stanmoore1** [comment]
Could also be a unsupported feature in the Kokkos comm.

**Comment by stanmoore1** [comment]
OK I believe 2cd95d1 fixed the segfault issues. Re-running regression tests.

**Comment by stanmoore1** [comment]
@valleymouth the `colloid` example now runs, but I'm still seeing a segfault in `srd/in.srd.mixture` that looks related to `AtomVecSphereKokkos`:

```
==22860== Invalid write of size 8
==22860==    at 0x472878: Kokkos::Impl::ParallelFor<AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP>, Kokkos::RangePolicy<Kokkos::OpenMP>, Kokkos::OpenMP>::execute() const [clone ._omp_fn.28] (atom_vec_sphere_kokkos.cpp:1588)
==22860==    by 0x4F4D8DE: GOMP_parallel (parallel.c:167)
==22860==    by 0x47DC6C: void Kokkos::parallel_for<AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP> >(unsigned long, AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP> const&, std::string const&) (Kokkos_OpenMP_Parallel.hpp:135)
==22860==    by 0x47502F: LAMMPS_NS::AtomVecSphereKokkos::unpack_exchange_kokkos(Kokkos::DualView<double**, Kokkos::LayoutRight, Kokkos::OpenMP, void>&, int, int, int, double, double, LAMMPS_NS::ExecutionSpace) (atom_vec_sphere_kokkos.cpp:1603)
==22860==    by 0x4A7B7E: void LAMMPS_NS::CommKokkos::exchange_device<Kokkos::OpenMP>() (comm_kokkos.cpp:673)
==22860==    by 0xA22B28: LAMMPS_NS::VerletKokkos::run(int) (verlet_kokkos.cpp:382)
==22860==    by 0x9E88A0: LAMMPS_NS::Run::command(int, char**) (run.cpp:183)
==22861== Invalid write of size 8
==22861==    at 0x472878: Kokkos::Impl::ParallelFor<AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP>, Kokkos::RangePolicy<Kokkos::OpenMP>, Kokkos::OpenMP>::execute() const [clone ._omp_fn.28] (atom_vec_sphere_kokkos.cpp:1588)
==22861==    by 0x4F4D8DE: GOMP_parallel (parallel.c:167)
==22861==    by 0x47DC6C: void Kokkos::parallel_for<AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP> >(unsigned long, AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP> const&, std::string const&) (Kokkos_OpenMP_Parallel.hpp:135)
==22861==    by 0x47502F: LAMMPS_NS::AtomVecSphereKokkos::unpack_exchange_kokkos(Kokkos::DualView<double**, Kokkos::LayoutRight, Kokkos::OpenMP, void>&, int, int, int, double, double, LAMMPS_NS::ExecutionSpace) (atom_vec_sphere_kokkos.cpp:1603)
==22861==    by 0x4A7B7E: void LAMMPS_NS::CommKokkos::exchange_device<Kokkos::OpenMP>() (comm_kokkos.cpp:673)
==22860==    by 0x6BD165: void LAMMPS_NS::Input::command_creator<LAMMPS_NS::Run>(LAMMPS_NS::LAMMPS*, int, char**) (input.cpp:873)
==22860==    by 0x6BB5A6: LAMMPS_NS::Input::execute_command() (input.cpp:856)
==22860==    by 0x6BCB16: LAMMPS_NS::Input::file() (input.cpp:243)
==22860==    by 0x6D1A55: main (main.cpp:64)
==22861==    by 0xA22B28: LAMMPS_NS::VerletKokkos::run(int) (verlet_kokkos.cpp:382)
==22861==    by 0x9E88A0: LAMMPS_NS::Run::command(int, char**) (run.cpp:183)
==22860==  Address 0x170 is not stack'd, malloc'd or (recently) free'd
==22860==
==22861==    by 0x6BD165: void LAMMPS_NS::Input::command_creator<LAMMPS_NS::Run>(LAMMPS_NS::LAMMPS*, int, char**) (input.cpp:873)
==22861==    by 0x6BB5A6: LAMMPS_NS::Input::execute_command() (input.cpp:856)
==22860==
==22860== Process terminating with default action of signal 11 (SIGSEGV)
==22860==  Access not within mapped region at address 0x170
==22861==    by 0x6BCB16: LAMMPS_NS::Input::file() (input.cpp:243)
==22861==    by 0x6D1A55: main (main.cpp:64)
==22861==  Address 0x160 is not stack'd, malloc'd or (recently) free'd
==22861==
==22860==    at 0x472878: Kokkos::Impl::ParallelFor<AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP>, Kokkos::RangePolicy<Kokkos::OpenMP>, Kokkos::OpenMP>::execute() const [clone ._omp_fn.28] (atom_vec_sphere_kokkos.cpp:1588)
==22860==    by 0x4F4D8DE: GOMP_parallel (parallel.c:167)
==22861==
==22861== Process terminating with default action of signal 11 (SIGSEGV)
==22861==  Access not within mapped region at address 0x160
==22860==    by 0x47DC6C: void Kokkos::parallel_for<AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP> >(unsigned long, AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP> const&, std::string const&) (Kokkos_OpenMP_Parallel.hpp:135)
==22860==    by 0x47502F: LAMMPS_NS::AtomVecSphereKokkos::unpack_exchange_kokkos(Kokkos::DualView<double**, Kokkos::LayoutRight, Kokkos::OpenMP, void>&, int, int, int, double, double, LAMMPS_NS::ExecutionSpace) (atom_vec_sphere_kokkos.cpp:1603)
==22861==    at 0x472878: Kokkos::Impl::ParallelFor<AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP>, Kokkos::RangePolicy<Kokkos::OpenMP>, Kokkos::OpenMP>::execute() const [clone ._omp_fn.28] (atom_vec_sphere_kokkos.cpp:1588)
==22861==    by 0x4F4D8DE: GOMP_parallel (parallel.c:167)
==22860==    by 0x4A7B7E: void LAMMPS_NS::CommKokkos::exchange_device<Kokkos::OpenMP>() (comm_kokkos.cpp:673)
==22860==    by 0xA22B28: LAMMPS_NS::VerletKokkos::run(int) (verlet_kokkos.cpp:382)
==22861==    by 0x47DC6C: void Kokkos::parallel_for<AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP> >(unsigned long, AtomVecSphereKokkos_UnpackExchangeFunctor<Kokkos::OpenMP> const&, std::string const&) (Kokkos_OpenMP_Parallel.hpp:135)
==22860==    by 0x9E88A0: LAMMPS_NS::Run::command(int, char**) (run.cpp:183)
==22860==    by 0x6BD165: void LAMMPS_NS::Input::command_creator<LAMMPS_NS::Run>(LAMMPS_NS::LAMMPS*, int, char**) (input.cpp:873)
==22861==    by 0x47502F: LAMMPS_NS::AtomVecSphereKokkos::unpack_exchange_kokkos(Kokkos::DualView<double**, Kokkos::LayoutRight, Kokkos::OpenMP, void>&, int, int, int, double, double, LAMMPS_NS::ExecutionSpace) (atom_vec_sphere_kokkos.cpp:1603)
==22860==    by 0x6BB5A6: LAMMPS_NS::Input::execute_command() (input.cpp:856)
==22860==    by 0x6BCB16: LAMMPS_NS::Input::file() (input.cpp:243)
==22861==    by 0x4A7B7E: void LAMMPS_NS::CommKokkos::exchange_device<Kokkos::OpenMP>() (comm_kokkos.cpp:673)
==22861==    by 0xA22B28: LAMMPS_NS::VerletKokkos::run(int) (verlet_kokkos.cpp:382)
==22860==    by 0x6D1A55: main (main.cpp:64)
==22860==  If you believe this happened as a result of a stack
==22860==  overflow in your program's main thread (unlikely but
==22860==  possible), you can try to increase the size of the
==22860==  main thread stack using the --main-stacksize= flag.
==22861==    by 0x9E88A0: LAMMPS_NS::Run::command(int, char**) (run.cpp:183)
==22861==    by 0x6BD165: void LAMMPS_NS::Input::command_creator<LAMMPS_NS::Run>(LAMMPS_NS::LAMMPS*, int, char**) (input.cpp:873)
==22860==  The main thread stack size used in this run was 10485760.
==22861==    by 0x6BB5A6: LAMMPS_NS::Input::execute_command() (input.cpp:856)
==22861==    by 0x6BCB16: LAMMPS_NS::Input::file() (input.cpp:243)
==22861==    by 0x6D1A55: main (main.cpp:64)
==22861==  If you believe this happened as a result of a stack
==22861==  overflow in your program's main thread (unlikely but
==22861==  possible), you can try to increase the size of the
==22861==  main thread stack using the --main-stacksize= flag.
==22861==  The main thread stack size used in this run was 10485760.
```

**Comment by stanmoore1** [comment]
I'm seeing the following tests failing: 

```
['examples/srd', 'srd.mixture']
['examples/pour', 'pour.2d.molecule']
['examples/pour', 'pour']
['examples/pour', 'pour.2d']
['examples/ASPHERE/poly', 'poly.mp']
['examples/ASPHERE/poly', 'poly']
```

**Comment by stanmoore1** [comment]
When `KOKKOS_DEBUG=yes`, the `ASPHERE `and `srd` examples are failing with `terminate called after throwing an instance of 'std::runtime_error' what():  View bounds error of view Neighbor::d_bincount ( 615817091 < 205 )`

The `pour` examples fail with `terminate called after throwing an instance of 'std::runtime_error' what():  View bounds error of view neighbor_history:firstflag ( 0 < 0 , 0 < 16 )`

**Comment by valleymouth** [comment]
@stanmoore1 Sorry for being offline for a while. I have just switched my development environment. I have seen that you solved a lot of issues already. I can resume my contributions tomorrow. Do you need any assistance?

**Comment by stanmoore1** [comment]
@valleymouth we are getting close, regression tests are running again. Could you take a look at `AtomVecSphereKokkos` and make sure each function has the correct sync/modify? I added a few that caused regression tests to fail, but there may be a few more that are missing. Thanks!

**Comment by stanmoore1** [comment]
All of the Kokkos OpenMP tests now pass (or fail with a meaningful error as expected). The `ASPHERE/poly` tests with Kokkos CUDA are failing with: `terminate called after throwing an instance of 'std::runtime_error' what():  Kokkos::Impl::ParallelFor< Cuda > requested too large team size.`. Looking into this.

**Comment by stanmoore1** [comment]
For `ASPHERE/poly`, if I reduce the binsize then the test runs fine with Kokkos CUDA, so it looks OK.

**Comment by stanmoore1** [comment]
Our internal Kokkos regression tests for both Kokkos CUDA and OpenMP now look good for this PR.

**Comment by valleymouth** [comment]
@stanmoore1 I reviewed  AtomVecSphereKokkos sync/modify calls and I actually spotted a few unnecessary calls. Not sure if you want to get rid of them now that you ended all regression tests.

**Comment by stanmoore1** [comment]
@valleymouth Thanks for checking. Yes please remove them in a commit, it is easy to rerun the tests. Thanks

**Comment by stanmoore1** [comment]
@valleymouth I updated the docs and restarted the regression tests. As soon as the tests pass, I think this PR is ready to be merged.

**Comment by akohlmey** [comment]
Removed conflicts due to the documentation refactoring currently happening in the master branch.

**Comment by akohlmey** [comment]
@stanmoore1 is this ready to be included or would it be better to wait until after the upcoming stable release? Please let me know by COB on Monday. From then on, only bugfix changes will be accepted until after the stable release.

**Comment by stanmoore1** [comment]
@valleymouth is currently working on some performance optimizations. @valleymouth what are your thoughts?

**Comment by valleymouth** [comment]
@stanmoore1 @akohlmey I think we can wait for the CommKokkos to be able to handle ghost velocities and exchange of fixes data, since it has a major impact in the overall performance.

**Comment by stanmoore1** [comment]
@valleymouth that sounds good to me. Thanks for the input.

**Comment by rbberger** [comment]
Did another merge with master to remove the conflicts

**Comment by akohlmey** [comment]
@stanmoore1 can you please comment on the status of this? is the needed change that is waiting for going to be available soon? that is, before the next planned stable release in mid/late november?

**Comment by stanmoore1** [comment]
@akohlmey I'd like to get this merged by next month if possible. I've been really busy but now finally have some time to work on this. @valleymouth was optimizing performance by keeping more data on the GPU but had some questions on path forward. I'm going to take a look and get back to him.

**Comment by stanmoore1** [comment]
Thanks @valleymouth. @akohlmey I'm running this through the Kokkos regression tests, as soon as those pass this should be ready to merge.

**Comment by stanmoore1** [comment]
For some reason the `coreshell` example hangs with the branch, looking into it.

**Comment by stanmoore1** [comment]
Regression tests now look, I believe this is ready to merge.

**Pull request merged**